Assertion failure when calling WKBundlePageHasLocalDataForURL() with a URL that contains a fragment identifier

jeffmapple opened the pull request:
#### 40c78192b01e6da5546b5ef53f6fa347fa1e2cd7
<pre>
Assertion failure when calling WKBundlePageHasLocalDataForURL() with a URL that contains a fragment identifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=252356">https://bugs.webkit.org/show_bug.cgi?id=252356</a>
rdar://105518815

Reviewed by Wenson Hsieh.

WKBundlePageHasLocalDataForURL() is the only caller of WebPage::hasLocalDataForURL(), which calls
DocumentLoader::subresource() with the URL. That function only expects URLs without fragment
identifiers, so remove any identifier first in WebPage::hasLocalDataForURL().

* Source/WebCore/loader/cache/MemoryCache.h:
Export MemoryCache::removeFragmentIdentifierIfNeeded() so it can be used from WebKit.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::hasLocalDataForURL):
Remove any fragment identifier from the URL before passing it to DocumentLoader::subresource().

Canonical link: <a href="https://commits.webkit.org/260371@main">https://commits.webkit.org/260371@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1f9033fb030f18c8d4ffee3a73005786049547b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17105 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40919 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117172 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116504 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111933 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18476 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8414 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100243 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113811 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13945 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97155 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41856 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95852 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28793 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83525 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10005 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30141 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10724 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7043 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16155 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49731 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7178 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12303 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

whsieh reviewed (approved):
~Instead of exposing and using the `MemoryCache` helper, could we just call `.removeFragmentIdentifier()`?~

(Seems like the answer is no, for consistency with logic in `CachedResourceLoader`)

jeffmapple committed:
Assertion failure when calling WKBundlePageHasLocalDataForURL() with a URL that contains a fragment identifier
https://bugs.webkit.org/show_bug.cgi?id=252356
rdar://105518815

Reviewed by Wenson Hsieh.

WKBundlePageHasLocalDataForURL() is the only caller of WebPage::hasLocalDataForURL(), which calls
DocumentLoader::subresource() with the URL. That function only expects URLs without fragment
identifiers, so remove any identifier first in WebPage::hasLocalDataForURL().

* Source/WebCore/loader/cache/MemoryCache.h:
Export MemoryCache::removeFragmentIdentifierIfNeeded() so it can be used from WebKit.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::hasLocalDataForURL):
Remove any fragment identifier from the URL before passing it to DocumentLoader::subresource().

Canonical link: https://commits.webkit.org/260371@main
